packages/@aws-cdk/core/lib/private/refs.ts

```diff
@@ -0,0 +1,276 @@
+import { Construct } from '../construct-compat';
+import { Reference } from '../reference';
+import { IResolvable } from '../resolvable';
+import { Stack } from '../stack';
+import { Token } from '../token';
+import { CfnReference } from './cfn-reference';
+import { Intrinsic } from './intrinsic';
+import { findTokens } from './resolve';
+import { makeUniqueId } from './uniqueid';
+
+/**
+ * This is called from the App level to resolve all references defined. Each
+ * reference is resolved based on it's consumption context.
+ */
+export function resolveReferences(scope: Construct) {
+  const edges = findAllReferences(scope);
+
+  for (const { source, value } of edges) {
+    const consumer = Stack.of(source);
+
+    // skip if we already have a value for this consumer
+    if (value.hasValueForStack(consumer)) {
+      continue;
+    }
+
+    // resolve the value in the context of the consumer.
+    const resolved = resolveValue(consumer, value);
+    value.assignValueForStack(consumer, resolved);
+  }
+}
+
+/**
+ * Resolves the value for `reference` in the context of `consumer`.
+ */
+function resolveValue(consumer: Stack, reference: CfnReference): IResolvable {
+  const producer = Stack.of(reference.target);
+
+  // produce and consumer stacks are the same, we can just return the value itself.
+  if (producer === consumer) {
+    return reference;
+  }
+
+  // unsupported: stacks from different apps
+  if (producer.node.root !== consumer.node.root) {
+    throw new Error('Cannot reference across apps. Consuming and producing stacks must be defined within the same CDK app.');
+  }
+
+  // unsupported: stacks are not in the same environment
+  if (producer.environment !== consumer.environment) {
+    throw new Error(
+      `Stack "${consumer.node.path}" cannot consume a cross reference from stack "${producer.node.path}". ` +
+      'Cross stack references are only supported for stacks deployed to the same environment or between nested stacks and their parent stack');
+  }
+
+  // ----------------------------------------------------------------------
+  // consumer is nested in the producer (directly or indirectly)
+  // ----------------------------------------------------------------------
+
+  // if the consumer is a child of the producer, wire the reference through a
+  // CloudFormation parameter on the consumer and resolve recursively.
+  if (isParent(producer, consumer)) {
+    const parameterValue = createNestedStackParameter(consumer, reference);
+    return resolveValue(consumer, parameterValue);
+  }
+
+  // ----------------------------------------------------------------------
+  // producer is a nested stack
+  // ----------------------------------------------------------------------
+
+  // if the producer is nested, always publish the value through a
+  // cloudformation output and resolve recursively with the Fn::GetAtt
+  // of the output in the parent stack.
+
+  // one might ask, if the consumer is not a parent of the producer,
```

Comment:
The problem would be how to synthesize the import side. We will need to somehow propagate the nested stack’s name to the consumer, so we might as well just propagate the actual value.

Reply:
There is one caveat to this, when the consumer stack and the producer stack are not in the same nested stack group, we must use export/import. We can create the export on the nested stack or on the root stack. Stacks have output/export limits,  If all are submitted on the root stack we are `wasting` outputs/export on the root stack. (disclaimer: the hard limit for outputs is 150 so one may argue this does not really matter)

Reply:
We will still need to somehow export the nested stack’s name from the top level stack, so this still uses at least one export for each nested stack and rather complicates the scheme so I prefer not to prematurely optimize this. We can introduce this later if there is a strong use case, right?

Reply:
Wouldn't it be a breaking change?  Exports can't be moved.

Reply:
Yes, you are correct. It will be a breaking change, and we can introduce under a feature flag if this will ever be an issue. 

Do you think it's worth the added complexity (and bugs that come with it)?

Reply:
Based on intuition alone, I strongly prefer the "all communication with the outside world goes through the top-level stack" version.

Feels more manageable/locally correct/verifiable... otherwise I'm getting visions of global variables.

Now, this might all be superstition and it's not really based on arguments. As I said it's just intuition.

Reply:
This is how it’s implemented. The value is outputted from the nested stack and will propagate all the way to the top before it is exported to the outside world.